File: runtool.py
#!/usr/bin/python

#author:clearluo
#create date:  2017-05-26
#modify date:2017-05-26

import os
import sys
import datetime
import time
#*/1 * * * * /usr/bin/python /home/didong/go/runtool.py monitor >> /home/didong/go/logs/crontab.log 2>&1
os.environ["GOPATH"] = "/home/didong/go"
os.environ["GOLOG"] = os.environ["GOPATH"] + "/logs"
os.environ["GOSRC"] = os.environ["GOPATH"] + "/src/didong"
os.environ["GOBIN"] = os.environ["GOPATH"] + "/bin"
os.environ["BIN"]="didong-backend"
curTime = "["+datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')+"] "
outPut = os.popen('ps -ef | grep $BIN | grep -v grep | awk \'{print $2}\'')
outPutStr = outPut.read()
curPid = 0
if outPutStr != "":
	curPid = int(outPutStr)
os.environ["CURPID"] = str(curPid)

if os.system('cd $GOPATH'):
	exit()

if len(sys.argv)!=2:
	print(curTime+sys.argv[0]+" start | stop | restart | status | monitor")
	exit()

def start():
	if os.system('$GOBIN/$BIN >> $GOLOG/temp.log 2>&1 &') == 0:
		time.sleep(3)
		newOutPut = os.popen('ps -ef | grep $BIN | grep -v grep | awk \'{print $2}\'')
		newOutPutStr = newOutPut.read()
		newCurPid = 0
		if newOutPutStr != "":
			newCurPid = int(newOutPutStr)
			print(curTime+"The program start success, PID:",newCurPid)
		else:
			print(curTime+"The program start failure 1.")
	else:
			print(curTime+"The program start failure 2.")
def stop():
	if os.system('kill -9 $CURPID') == 0:
		print(curTime+"The program stop success, PID:",curPid)

cmd = sys.argv[1]
if cmd == "start":
	if curPid == 0:
		start()
	else:
		print(curTime+"The program already running, PID:",curPid)
elif cmd == "stop":
	if curPid == 0:
		print(curTime+"The program is not running.")
	else:
		stop()
elif cmd == "restart":
	if curPid == 0:
		start()
	else:
		stop()
		time.sleep(3)
		start()
elif cmd == "status":
	if curPid == 0:
		print(curTime+"The program is not running.")
	else:
		print(curTime+"The current program PID:",curPid)
elif cmd == "monitor":
	if curPid == 0:
		print(curTime+"Monitor program, program dead, start program.")
		start()
	else:
		print(curTime+"Monitor program, PID:",curPid)
else:
	print(curTime+sys.argv[0]+" start | stop | restart | status | monitor")
	exit()

File: runtool.go
//#!/usr/bin/env gorun
package main

/**
 * go get github.com/erning/gorun
 */
import (
	"bytes"
	"fmt"
	"log"
	"os"
	"os/exec"
	"time"
)

type Program struct {
	GoPath string
	GoPro  string
	GoBin  string
	GoLog  string
	Exe    string
	CurPid string
}

func init() {
	log.SetFlags(log.LstdFlags)
	time.Local = time.FixedZone("CST", 3600*8)
}

func main() {
	program := Program{
		GoPath: os.Getenv("GOPATH"),
		GoPro:  os.Getenv("GOPATH") + "/src/didong",
		GoBin:  os.Getenv("GOPATH") + "/bin",
		GoLog:  os.Getenv("GOPATH") + "/logs",
		Exe:    "didong-backend",
	}
	if len(os.Args) != 2 {
		printHelp()
		os.Exit(0)
	}
	program.CurPid = program.GetPid()
	switch os.Args[1] {
	case "start":
		if "" == program.CurPid {
			program.Start()
		} else {
			log.Println("The program already running, PID:", program.CurPid)
		}
	case "stop":
		if "" == program.CurPid {
			log.Println("The program is not running.")
		} else {
			program.Stop()
		}
	case "restart":
		if "" == program.CurPid {
			program.Start()
		} else {
			program.Stop()
			time.Sleep(time.Second * 3)
			program.Start()
		}
	case "status":
		if "" == program.CurPid {
			log.Println("The program is not running.")
		} else {
			log.Println("The current program PID:", program.CurPid)
		}
	case "monitor":
		if "" == program.CurPid {
			log.Println("Monitor program, program dead, start program.")
			program.Start()
		} else {
			log.Println("Monitor program, PID:", program.CurPid)
		}
	default:
		printHelp()
	}
}

func printHelp() {
	fmt.Println("./xx start | stop | restart | status | monitor")
}

func GetCurTime() string {
	timestamp := time.Now().Unix()
	tm := time.Unix(timestamp, 0)
	return tm.Format("2006-01-02 15:04:05")
}

func TrimEnd(result []byte) string {
	if len(result) > 2 {
		result = append(result[0:0], result[0:len(result)-1]...)
	}
	return string(result)
}
func ExecShell(cmd string) (string, error) {
	ret, err := exec.Command("bash", "-c", cmd).Output()
	if err != nil {
		log.Println(err)
		os.Exit(0)
		return "", err
	}
	return TrimEnd(ret), nil
}
func (p Program) GetPid() string {
	cmd := bytes.Buffer{}
	cmd.WriteString(`ps -ef | grep `)
	cmd.WriteString(p.Exe)
	cmd.WriteString(` | grep -v grep | awk '{print $2}'`)
	ret, _ := ExecShell(cmd.String())
	return ret
}

func (p *Program) Start() {
	if p.GetPid() != "" {
		log.Println("The program already running, PID:", p.CurPid)
		os.Exit(0)
	}
	cmd := p.GoBin + "/" + p.Exe + " >> " + p.GoLog + "/temp.log 2>&1 &"
	ExecShell(cmd)
	time.Sleep(time.Second * 3)
	p.CurPid = p.GetPid()
	if "" == p.CurPid {
		log.Println("The program start failure.")
	} else {
		log.Println("The program start success, PID:", p.CurPid)
	}
}
func (p Program) Stop() {
	ExecShell("kill -9 " + p.CurPid)
	log.Println("The program stop success, PID:", p.CurPid)
}
